# 回帰

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from IPython.core.display import display

In [2]:
from sklearn.datasets import load_boston
dataset =  load_boston()

In [3]:
X =  pd.DataFrame( dataset.data,columns=dataset.feature_names)

In [4]:
X.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [5]:
y =  pd.Series(dataset.target,name='y')
y.head

<bound method NDFrame.head of 0      24.0
1      21.6
2      34.7
3      33.4
4      36.2
5      28.7
6      22.9
7      27.1
8      16.5
9      18.9
10     15.0
11     18.9
12     21.7
13     20.4
14     18.2
15     19.9
16     23.1
17     17.5
18     20.2
19     18.2
20     13.6
21     19.6
22     15.2
23     14.5
24     15.6
25     13.9
26     16.6
27     14.8
28     18.4
29     21.0
       ... 
476    16.7
477    12.0
478    14.6
479    21.4
480    23.0
481    23.7
482    25.0
483    21.8
484    20.6
485    21.2
486    19.1
487    20.6
488    15.2
489     7.0
490     8.1
491    13.6
492    20.1
493    21.8
494    24.5
495    23.1
496    19.7
497    18.3
498    21.2
499    17.5
500    16.8
501    22.4
502    20.6
503    23.9
504    22.0
505    11.9
Name: y, Length: 506, dtype: float64>

In [6]:
print(y.describe())

count    506.000000
mean      22.532806
std        9.197104
min        5.000000
25%       17.025000
50%       21.200000
75%       25.000000
max       50.000000
Name: y, dtype: float64


In [7]:
display(X.join(y).head(5)) # 先頭にジョインされる

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT     y  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [8]:
from sklearn.ensemble import GradientBoostingRegressor
est = GradientBoostingRegressor( n_estimators=100,
                               max_depth=3,
                               random_state=0)
est.fit( X,y)

GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
             learning_rate=0.1, loss='ls', max_depth=3, max_features=None,
             max_leaf_nodes=None, min_impurity_decrease=0.0,
             min_impurity_split=None, min_samples_leaf=1,
             min_samples_split=2, min_weight_fraction_leaf=0.0,
             n_estimators=100, n_iter_no_change=None, presort='auto',
             random_state=0, subsample=1.0, tol=0.0001,
             validation_fraction=0.1, verbose=0, warm_start=False)

In [9]:
# 回帰のスコアを測るのにr2スコアを使う。
rom sklearn.metrics import r2_score
y_pred = est.predict(X)
r2 =  r2_score( y,y_pred)
print('R2 socre %.3f'% r2)

R2 socre 0.976


In [10]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.20,random_state= 1)

In [11]:
est1 = GradientBoostingRegressor(max_depth=3,random_state=1)
est1.fit(X_train,y_train)
est2 = GradientBoostingRegressor(max_depth=10,random_state=1) 
est2.fit(X_train,y_train)

GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
             learning_rate=0.1, loss='ls', max_depth=10, max_features=None,
             max_leaf_nodes=None, min_impurity_decrease=0.0,
             min_impurity_split=None, min_samples_leaf=1,
             min_samples_split=2, min_weight_fraction_leaf=0.0,
             n_estimators=100, n_iter_no_change=None, presort='auto',
             random_state=1, subsample=1.0, tol=0.0001,
             validation_fraction=0.1, verbose=0, warm_start=False)

In [12]:
r2_est1_train = r2_score( y_train,est1.predict(X_train))
r2_est2_train = r2_score( y_train,est2.predict(X_train)) 
# for test data
r2_est1_test = r2_score(y_test,est1.predict(X_test))
r2_est2_test = r2_score(y_test,est2.predict(X_test))


In [13]:
# 性能指標の表示
print('Train Score(est1,est2):(%.3f,%.3f)'% (r2_est1_train, r2_est2_train))
print('Test Score(est1,est2):(%.3f,%.3f)'% (r2_est1_test, r2_est2_test))

Train Score(est1,est2):(0.978,1.000)
Test Score(est1,est2):(0.917,0.847)


In [17]:
print('--------------------------------------------')
print('X shape: (%i,%i)' %X.shape)
print('--------------------------------------------')
display(X.join(y).head())

--------------------------------------------
X shape: (506,13)
--------------------------------------------


CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT     y  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [37]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression, Ridge,Lasso
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor,GradientBoostingRegressor
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
# リッジ回帰ではL2iペナルティによって係数を縮小して過学習を抑える。

In [30]:
# ホールドアウトのためデータを訓練とテストに分割
X_train,X_test,y_train,y_test = train_test_split(X,
                                                 y,
                                                 test_size=0.20,
                                                 random_state=1)

In [38]:
pipelines =  {
    'ols':Pipeline([( 'scl',StandardScaler()),('est',LinearRegression())])
    ,'ridge1':Pipeline([('scl',StandardScaler()),('est',Ridge(alpha=1.0))])
    ,'ridge2':Pipeline([('scl',StandardScaler()),('est',Ridge(alpha=20.0))])
    ,'lasso':Pipeline([('scl',StandardScaler()),('est',Lasso(alpha=1.0))])
    , 'tree': Pipeline([('scl',StandardScaler()),('est',DecisionTreeRegressor())])
    ,'rf': Pipeline([('scl',StandardScaler()), ('est',RandomForestRegressor(random_state=0))])
    , 'gbr1': Pipeline([('scl',StandardScaler()),('est',GradientBoostingRegressor())])
    , 'gbr2': Pipeline([('scl',StandardScaler()),('est',GradientBoostingRegressor(n_estimators=250,))])    
}

In [39]:
scores =  {}
for pipe_name, est in pipelines.items():
    est.fit( X_train,y_train)
    scores[('train',pipe_name)]=r2_score(y_train, est.predict(X_train))
    scores[('test',pipe_name)]=r2_score(y_test, est.predict(X_test))
display( pd.Series(scores).unstack())

/Users/humu/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


gbr1      gbr2     lasso       ols        rf    ridge1    ridge2  \
test   0.926425  0.924903  0.654651  0.763417  0.884060  0.763404  0.758157   
train  0.977915  0.994840  0.651037  0.729359  0.977035  0.729336  0.725291   

           tree  
test   0.813683  
train  1.000000

In [40]:
pipelines

{'ols': Pipeline(memory=None,
      steps=[('scl', StandardScaler(copy=True, with_mean=True, with_std=True)), ('est', LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
          normalize=False))]), 'ridge1': Pipeline(memory=None,
      steps=[('scl', StandardScaler(copy=True, with_mean=True, with_std=True)), ('est', Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
    normalize=False, random_state=None, solver='auto', tol=0.001))]), 'ridge2': Pipeline(memory=None,
      steps=[('scl', StandardScaler(copy=True, with_mean=True, with_std=True)), ('est', Ridge(alpha=20.0, copy_X=True, fit_intercept=True, max_iter=None,
    normalize=False, random_state=None, solver='auto', tol=0.001))]), 'lasso': Pipeline(memory=None,
      steps=[('scl', StandardScaler(copy=True, with_mean=True, with_std=True)), ('est', Lasso(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=1000,
    normalize=False, positive=False, precompute=False, random_state=None,
    selection='cycl

In [41]:
# アルゴリズム→step数→係数→の合計　で合計係数を表示する。
print('OLS coefficient total:%.6f'%(np.absolute(pipelines['ols'].named_steps['est'].coef_).sum()))
print('Ridge coefficient total:%.6f'%(np.absolute(pipelines['ridge1'].named_steps['est'].coef_).sum()))
print('Ridge coefficient total:%.6f'%(np.absolute(pipelines['ridge2'].named_steps['est'].coef_).sum()))

OLS coefficient total:22.063408
Ridge coefficient total:21.710242
Ridge coefficient total:18.013609


In [42]:
from sklearn.model_selection import cross_val_score

scores = {}
for pipe_name,est in pipelines.items():
    cv_results = cross_val_score(est,X,y,cv=5,scoring='r2') # cvは繰り返す数
    print('----------')
    print('algorithm:', pipe_name)
    print('cv_results:', cv_results)
    print('avg +- std_dev', cv_results.mean(),'+-', cv_results.std())
    

----------
algorithm: ols
cv_results: [ 0.63919994  0.71386698  0.58702344  0.07923081 -0.25294154]
avg +- std_dev 0.3532759243958822 +- 0.376567839332623
----------
algorithm: ridge1
cv_results: [ 0.64344111  0.71648023  0.58788768  0.08218971 -0.23681375]
avg +- std_dev 0.35863699557121664 +- 0.37221115867544025
----------
algorithm: ridge2
cv_results: [ 0.69338962  0.74221138  0.59160501  0.12868283 -0.04192937]
avg +- std_dev 0.4227918951161508 +- 0.31818730396014017
----------
algorithm: lasso
cv_results: [ 0.72326572  0.67913186  0.32264318  0.1588301  -0.21487988]
avg +- std_dev 0.33379819796717003 +- 0.34719625237343843
----------
algorithm: tree
cv_results: [ 0.64050598  0.41103213  0.62952512  0.40417978 -0.82965793]
avg +- std_dev 0.2511170153082703 +- 0.5498894428234987
----------
algorithm: rf
cv_results: [0.78532278 0.84330892 0.72098833 0.45610161 0.18910517]
avg +- std_dev 0.5989653618688409 +- 0.24401553002201648
----------
algorithm: gbr1
cv_results: [0.7839646  0.855

# 分類

In [43]:
# Load and return the breast cancer wisconsin dataset (classification).
# The breast cancer dataset is a classic and very easy binary classification dataset.
import pandas as pd
from sklearn.datasets import load_breast_cancer

# set data by role
dataset = load_breast_cancer()
X = pd.DataFrame(dataset.data, columns=dataset.feature_names)
y = pd.Series(dataset.target, name='y')

# check the shape
print('--------------------------------------------------')
print('X shape: (%i,%i)' %X.shape)
print('--------------------------------------------------')
print(y.value_counts())
print('--------------------------------------------------')
print('y=0 means Marignant(悪性),y=1 means Benign(良性):')
print('--------------------------------------------------')
X.join(y).head()

--------------------------------------------------
X shape: (569,30)
--------------------------------------------------
1    357
0    212
Name: y, dtype: int64
--------------------------------------------------
y=0 means Marignant(悪性),y=1 means Benign(良性):
--------------------------------------------------


mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  y  
0          0.4601                  0.11890  0  
1          0.2750                  0.08902  0  
2          0.3613                  0.08758  0  
3          0.6638                  0.17300  0  
4          0.2364                  0.07678  0  

[5 rows x 31 columns]

次に評価指標についてです。回帰モデルの評価にはR2スコアなどがありました。分類モデルの評価の場合、最もわかりやすいものが正解率（Accuracy）です。正解率は予測対象データ（スコアリングデータ）の全件をNとした時、正例を正しく正例と分類できた数TP、負例を正しく負例と分類できた数TFとすると、 以下の数式で表せます。<br><center>Accuracy = (TP + TF) / N</center><br>つまり、正例・負例を問わず正しく分類されたサンプル数の割合です。sklearnでは正解データと予測データを与えると、Accuracyを計算してくれる関数が存在します。以下で呼び出しておきます。

In [44]:
# 分類モデルの評価指標計算のための関数の読込
from sklearn.metrics import accuracy_score # 分離用

In [45]:
# import libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

# Holdout
X_train,X_test,y_train,y_test = train_test_split(X,
                                                 y,
                                                 test_size=0.20,
                                                 random_state=1)
# set pipelines for two different algorithms

In [47]:
pipelines =  {
    'Knn':Pipeline([('scl',StandardScaler()),('est',KNeighborsClassifier())]),
    'logistic': Pipeline([('scl',StandardScaler()),('est',LogisticRegression(random_state=1))])    
}
for pipe_name, pipeline in pipelines.items():
    pipeline.fit(X_train,y_train)
    print(pipe_name, ': Fitting Done')

Knn : Fitting Done
logistic : Fitting Done


/Users/humu/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [48]:
pipelines['logistic'].predict(X_train)[:10]

array([0, 0, 1, 1, 1, 1, 1, 0, 0, 1])

In [49]:
pipelines['logistic'].predict_proba(X_train)[:10] # 0,1のprobability

array([[9.99441634e-01, 5.58366357e-04],
       [9.99962234e-01, 3.77655428e-05],
       [1.12299304e-06, 9.99998877e-01],
       [4.10316174e-04, 9.99589684e-01],
       [1.83025794e-04, 9.99816974e-01],
       [7.88575120e-03, 9.92114249e-01],
       [6.74588243e-02, 9.32541176e-01],
       [9.99956558e-01, 4.34422063e-05],
       [9.99999212e-01, 7.88384042e-07],
       [2.32654465e-02, 9.76734554e-01]])

In [50]:
scores = {}
for pipe_name, pipeline in pipelines.items():
    scores[(pipe_name,'train')] = accuracy_score(y_train, pipeline.predict(X_train))
    scores[(pipe_name,'test')] = accuracy_score(y_test, pipeline.predict(X_test))

pd.Series(scores).unstack()

test     train
Knn       0.956140  0.982418
logistic  0.982456  0.991209

In [51]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC, LinearSVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# holdout
X_train,X_test,y_train,y_test=train_test_split(X,
                                               y,
                                               test_size=0.3,
                                               random_state=1)
# set pipelines for different algorithms
pipelines = {
    'knn':
        Pipeline([('scl',StandardScaler()),
                  ('est',KNeighborsClassifier())]),
    'logistic':
        Pipeline([('scl',StandardScaler()),
                  ('est',LogisticRegression(random_state=1))]),
    'rsvc':
        Pipeline([('scl',StandardScaler()),
                  ('est',SVC(C=1.0, kernel='rbf', class_weight='balanced', random_state=1))]),
    'lsvc':
        Pipeline([('scl',StandardScaler()),
                  ('est',LinearSVC(C=1.0, class_weight='balanced', random_state=1))]),
    'tree':
        Pipeline([('scl',StandardScaler()),
                  ('est',DecisionTreeClassifier(random_state=1))]),
    'rf':
        Pipeline([('scl',StandardScaler()),
                  ('est',RandomForestClassifier(random_state=1))]),
    'gb':
        Pipeline([('scl',StandardScaler()),
                  ('est',GradientBoostingClassifier(random_state=1))]),
    'mlp':
        Pipeline([('scl',StandardScaler()),
                  ('est',MLPClassifier(hidden_layer_sizes=(3,3),
                                       max_iter=1000,
                                       random_state=1))])
}
# fit & evaluation
scores = {}
for pipe_name, pipeline in pipelines.items():
    pipeline.fit(X_train, y_train)
    scores[(pipe_name,'train')] = accuracy_score(y_train, pipeline.predict(X_train))
    scores[(pipe_name,'test')] = accuracy_score(y_test, pipeline.predict(X_test))

pd.Series(scores).unstack()

/Users/humu/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/humu/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


test     train
gb        0.964912  1.000000
knn       0.959064  0.979899
logistic  0.970760  0.989950
lsvc      0.959064  0.987437
mlp       0.959064  0.989950
rf        0.935673  1.000000
rsvc      0.964912  0.984925
tree      0.929825  1.000000

<b>[確認してみよう]</b>アルゴリズムのハイパーパラメータの数値を様々に組み合わせ結果を確認してみよう。
- サポートベクターのclass_weight='balanced'を外したらどうなるか？
- サポートベクターのカーネルの種類を変えたらどうなるか？
- hidden_layer_sizes=(3,3)を(3,3,3)にしたらどうなるか？ etc.

In [54]:
pipelines = {
    'rsvc1':
        Pipeline([('scl',StandardScaler()),
                  ('est',SVC(C=1.0, kernel='rbf', class_weight='balanced', random_state=1))]),
    'rsvc2':
            Pipeline([('scl',StandardScaler()),
                      ('est',SVC(C=1.0, kernel='rbf', random_state=1))]),
    'lsvc1':
        Pipeline([('scl',StandardScaler()),
                  ('est',LinearSVC(C=1.0, class_weight='balanced', random_state=1))]),
    
    'lsvc2':
        Pipeline([('scl',StandardScaler()),
                  ('est',LinearSVC(C=1.0, random_state=1))]),
    'mlp1':
        Pipeline([('scl',StandardScaler()),
                  ('est',MLPClassifier(hidden_layer_sizes=(3,3),
                                       max_iter=1000,
                                       random_state=1))]),
      'mlp2':
        Pipeline([('scl',StandardScaler()),
                  ('est',MLPClassifier(hidden_layer_sizes=(3,3,3),
                                       max_iter=1000,
                                       random_state=1))])
}
# fit & evaluation
scores = {}
for pipe_name, pipeline in pipelines.items():
    pipeline.fit(X_train, y_train)
    scores[(pipe_name,'train')] = accuracy_score(y_train, pipeline.predict(X_train))
    scores[(pipe_name,'test')] = accuracy_score(y_test, pipeline.predict(X_test))

pd.Series(scores).unstack()

test     train
lsvc1  0.959064  0.987437
lsvc2  0.959064  0.992462
mlp1   0.959064  0.989950
mlp2   0.935673  0.994975
rsvc1  0.964912  0.984925
rsvc2  0.970760  0.989950

# 前処理

In [64]:
import pandas as pd
df = pd.read_csv('../python/data/av_loan_u6lujuX_CVtuZ9i.csv',header=0)
df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [65]:
df.shape

(614, 13)

In [69]:
X =  df.iloc[:,:-1] # 一番前から最終列の以前を特徴量に入れる。
X = X.drop('Loan_ID',axis=1) # 1列目はID情報のため特徴量から削除
y = df.iloc[:,-1]            # 最終列を正解データ

# check the shape
print('X shape: (%i,%i)' %X.shape)

# ローン審査でNOとなったサンプルを1（正例）へ変換
class_mapping = {'N':1, 'Y':0}
y = y.map(class_mapping)
print('--------------------')
print(y.value_counts())
X.join(y).head()


X shape: (614,11)
--------------------
0    422
1    192
Name: Loan_Status, dtype: int64


Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
0   Male      No          0      Graduate            No             5849   
1   Male     Yes          1      Graduate            No             4583   
2   Male     Yes          0      Graduate           Yes             3000   
3   Male     Yes          0  Not Graduate            No             2583   
4   Male      No          0      Graduate            No             6000   

   CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                0.0         NaN             360.0             1.0   
1             1508.0       128.0             360.0             1.0   
2                0.0        66.0             360.0             1.0   
3             2358.0       120.0             360.0             1.0   
4                0.0       141.0             360.0             1.0   

  Property_Area  Loan_Status  
0         Urban            0  
1         Rural            1  
2         Urban            0  
3         Urban            0  
4         Urban            0

In [70]:
pd.options.display.max_columns = 50 # データフレームの表示列数を50に拡張するよう、オプション変更をしておきます。

さて上記データを使って、以下のような前処理方法を学びます。
- <b>欠損値補完</b>：LoanAmountの1行目の欠損値をLoanAmount列の平均値で置き換える
- <b>one-hotエンコーディング</b>：Genderなどのカテゴリ変数を0/1のバイナリ変数に変換する

本講座では一連の前処理を、以下のような順番で統一します。
1. まず、<b>カテゴリ変数の欠損値</b>をone-hotエンコードをして解決し、
1. 次に、<b>数値変数の欠損値</b>を平均値で置き換え解決する。

それではone-hotエンコードの実施です。dummy_na=Trueとすると、欠損であったという情報も含めone-hot化されます。

In [75]:
#  カテゴリー変数　を選んでくる。 ２分類値
ohe_columns = ['Dependents',
               'Gender',
               'Married',
               'Education',
               'Self_Employed',
               'Property_Area']
X_new = pd.get_dummies(X, # one_hotエンコーディング
                                              dummy_na=True, # nanを埋める。
                                            columns=ohe_columns)
display(X_new.head())

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History  Dependents_0  Dependents_1  Dependents_2  Dependents_3+  \
0             1.0             1             0             0              0   
1             1.0             0             1             0              0   
2             1.0             1             0             0              0   
3             1.0             1             0             0              0   
4             1.0             1             0             0              0   

   Dependents_nan  Gender_Female  Gender_Male  Gender_nan  Married_No  \
0               0              0            1           0           1   
1               0              0            1           0           0   
2               0              0            1           0           0   
3               0              0            1           0           0   
4               0              0            1           0           1   

   Married_Yes  Married_nan  Education_Graduate  Education_Not Graduate  \
0            0            0                   1                       0   
1            1            0                   1                       0   
2            1            0                   1                       0   
3            1            0                   0                       1   
4            0            0                   1                       0   

   Education_nan  Self_Employed_No  Self_Employed_Yes  Self_Employed_nan  \
0              0                 1                  0                  0   
1              0                 1                  0                  0   
2              0                 0                  1                  0   
3              0                 1                  0                  0   
4              0                 1                  0                  0   

   Property_Area_Rural  Property_Area_Semiurban  Property_Area_Urban  \
0                    0                        0                    1   
1                    1                        0                    0   
2                    0                        0                    1   
3                    0                        0                    1   
4                    0                        0                    1   

   Property_Area_nan  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0

In [77]:
X_new.describe()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
count       614.000000         614.000000  592.000000         600.00000   
mean       5403.459283        1621.245798  146.412162         342.00000   
std        6109.041673        2926.248369   85.587325          65.12041   
min         150.000000           0.000000    9.000000          12.00000   
25%        2877.500000           0.000000  100.000000         360.00000   
50%        3812.500000        1188.500000  128.000000         360.00000   
75%        5795.000000        2297.250000  168.000000         360.00000   
max       81000.000000       41667.000000  700.000000         480.00000   

       Credit_History  Dependents_0  Dependents_1  Dependents_2  \
count      564.000000    614.000000    614.000000    614.000000   
mean         0.842199      0.561889      0.166124      0.164495   
std          0.364878      0.496559      0.372495      0.371027   
min          0.000000      0.000000      0.000000      0.000000   
25%          1.000000      0.000000      0.000000      0.000000   
50%          1.000000      1.000000      0.000000      0.000000   
75%          1.000000      1.000000      0.000000      0.000000   
max          1.000000      1.000000      1.000000      1.000000   

       Dependents_3+  Dependents_nan  Gender_Female  Gender_Male  Gender_nan  \
count     614.000000      614.000000     614.000000   614.000000  614.000000   
mean        0.083062        0.024430       0.182410     0.796417    0.021173   
std         0.276201        0.154506       0.386497     0.402991    0.144077   
min         0.000000        0.000000       0.000000     0.000000    0.000000   
25%         0.000000        0.000000       0.000000     1.000000    0.000000   
50%         0.000000        0.000000       0.000000     1.000000    0.000000   
75%         0.000000        0.000000       0.000000     1.000000    0.000000   
max         1.000000        1.000000       1.000000     1.000000    1.000000   

       Married_No  Married_Yes  Married_nan  Education_Graduate  \
count  614.000000   614.000000   614.000000          614.000000   
mean     0.346906     0.648208     0.004886            0.781759   
std      0.476373     0.477919     0.069786            0.413389   
min      0.000000     0.000000     0.000000            0.000000   
25%      0.000000     0.000000     0.000000            1.000000   
50%      0.000000     1.000000     0.000000            1.000000   
75%      1.000000     1.000000     0.000000            1.000000   
max      1.000000     1.000000     1.000000            1.000000   

       Education_Not Graduate  Education_nan  Self_Employed_No  \
count              614.000000          614.0        614.000000   
mean                 0.218241            0.0          0.814332   
std                  0.413389            0.0          0.389155   
min                  0.000000            0.0          0.000000   
25%                  0.000000            0.0          1.000000   
50%                  0.000000            0.0          1.000000   
75%                  0.000000            0.0          1.000000   
max                  1.000000            0.0          1.000000   

       Self_Employed_Yes  Self_Employed_nan  Property_Area_Rural  \
count         614.000000         614.000000           614.000000   
mean            0.133550           0.052117             0.291531   
std             0.340446           0.222445             0.454838   
min             0.000000           0.000000             0.000000   
25%             0.000000           0.000000             0.000000   
50%             0.000000           0.000000             0.000000   
75%             0.000000           0.000000             1.000000   
max             1.000000           1.000000             1.000000   

       Property_Area_Semiurban  Property_Area_Urban  Property_Area_nan  
count               614.000000           614.000000              614.0  
mean                  0.379479             0.328

上記まででカテゴリ変数の数量化と欠損値対応は終了です。<br>次に数値変数の欠損を平均値で置き換えます。

それでは、連続変数の欠損値の平均値補完の実行です。`sklearn.impute`モジュールの`SimpleImputer`クラスを読み込みます。SimpleImputerクラスのメソッドtransfomrを適用することで、LoanAmountの欠損値（1行目など）を平均値（146.412162）に置換できます。

In [82]:
from sklearn.impute import SimpleImputer # impute で欠損値を入れる
imp = SimpleImputer() # 欠損値に平均値を入れる。
imp.fit(X_new)
X_new_columns =  X_new.columns.values
X_new = pd.DataFrame(imp.transform(X_new),columns=X_new_columns)
display(X_new.head())

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0           5849.0                0.0  146.412162             360.0   
1           4583.0             1508.0  128.000000             360.0   
2           3000.0                0.0   66.000000             360.0   
3           2583.0             2358.0  120.000000             360.0   
4           6000.0                0.0  141.000000             360.0   

   Credit_History  Dependents_0  Dependents_1  Dependents_2  Dependents_3+  \
0             1.0           1.0           0.0           0.0            0.0   
1             1.0           0.0           1.0           0.0            0.0   
2             1.0           1.0           0.0           0.0            0.0   
3             1.0           1.0           0.0           0.0            0.0   
4             1.0           1.0           0.0           0.0            0.0   

   Dependents_nan  Gender_Female  Gender_Male  Gender_nan  Married_No  \
0             0.0            0.0          1.0         0.0         1.0   
1             0.0            0.0          1.0         0.0         0.0   
2             0.0            0.0          1.0         0.0         0.0   
3             0.0            0.0          1.0         0.0         0.0   
4             0.0            0.0          1.0         0.0         1.0   

   Married_Yes  Married_nan  Education_Graduate  Education_Not Graduate  \
0          0.0          0.0                 1.0                     0.0   
1          1.0          0.0                 1.0                     0.0   
2          1.0          0.0                 1.0                     0.0   
3          1.0          0.0                 0.0                     1.0   
4          0.0          0.0                 1.0                     0.0   

   Education_nan  Self_Employed_No  Self_Employed_Yes  Self_Employed_nan  \
0            0.0               1.0                0.0                0.0   
1            0.0               1.0                0.0                0.0   
2            0.0               0.0                1.0                0.0   
3            0.0               1.0                0.0                0.0   
4            0.0               1.0                0.0                0.0   

   Property_Area_Rural  Property_Area_Semiurban  Property_Area_Urban  \
0                  0.0                      0.0                  1.0   
1                  1.0                      0.0                  0.0   
2                  0.0                      0.0                  1.0   
3                  0.0                      0.0                  1.0   
4                  0.0                      0.0                  1.0   

   Property_Area_nan  
0                0.0  
1                0.0  
2                0.0  
3                0.0  
4                0.0

以上で、カテゴリ変数の数量化と欠損対応、数値変数の欠損値対応は終了です。もし特徴量の次元数（列数）が小さければ、そのままアルゴリズムに投入して構いません。ただし、実務では数百・数千次元を超えることがしばしばありますので、その際は以下の次元圧縮を行います。

# 次元圧縮(RFE&PCA)

さて、特徴量が元の11次元が26次元まで増加しました。<br><b>ここではRFEを使って、予測に役立つと判断された上位10変数に絞り込むこととします。</b>

In [87]:
from sklearn.feature_selection import RFE
from sklearn.ensemble import RandomForestClassifier

In [89]:
selector =  RFE(estimator= RandomForestClassifier(n_estimators= 100,random_state= 0),
               n_features_to_select=10,
               step= 0.5)
selector.fit( X_new,y)
print('Done normally')

Done normally


RFEをfitすることで、26変数のうちどの変数を残すかが決定されました。<br><b>残された変数の確認は"support_"属性を呼び出すことで可能です。</b><br>Trueが採用された変数の場所を表しています。

In [90]:
print(selector.support_)

[ True  True  True  True  True  True  True False False False False False
 False  True False False False False False  True False False False  True
 False False]


fitまでで残すべき変数を決めることができたので、実際にデータの絞り込みをします。<br>`SimpleImputer`クラスの`transform`メソッドで列の絞り込みができます。

In [91]:
# 26次元を10次元を圧縮
X_new_selected=selector.transform(X_new)
X_new_selected=pd.DataFrame(X_new_selected,
                            columns=X_new_columns[selector.support_])
print('X shape after RFE:', X_new_selected.shape)
print('---------------------------------------')
print(X_new_selected.dtypes)
X_new_selected.head()

X shape after RFE: (614, 10)
---------------------------------------
ApplicantIncome            float64
CoapplicantIncome          float64
LoanAmount                 float64
Loan_Amount_Term           float64
Credit_History             float64
Dependents_0               float64
Dependents_1               float64
Married_No                 float64
Self_Employed_No           float64
Property_Area_Semiurban    float64
dtype: object


ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0           5849.0                0.0  146.412162             360.0   
1           4583.0             1508.0  128.000000             360.0   
2           3000.0                0.0   66.000000             360.0   
3           2583.0             2358.0  120.000000             360.0   
4           6000.0                0.0  141.000000             360.0   

   Credit_History  Dependents_0  Dependents_1  Married_No  Self_Employed_No  \
0             1.0           1.0           0.0         1.0               1.0   
1             1.0           0.0           1.0         0.0               1.0   
2             1.0           1.0           0.0         0.0               0.0   
3             1.0           1.0           0.0         0.0               1.0   
4             1.0           1.0           0.0         1.0               1.0   

   Property_Area_Semiurban  
0                      0.0  
1                      0.0  
2                      0.0  
3                      0.0  
4                      0.0

最後に、PCAによる次元圧縮の方法を確認します。<br>X_new（RFE前の26次元）を対象にPCAで特徴抽出したい場合は、以下の通りです。

In [93]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import f1_score

# パイプラインにPCAを埋め込めば自動的に次元圧縮してくれる
clf = Pipeline([('scl', StandardScaler()),
                ('reduct', PCA(n_components=10,random_state=1)),
                ('clf', GradientBoostingClassifier(random_state=1))])

# 学習時に自動的にPCA処理が施される
clf.fit(X_new, y)
print('Normally done')

Normally done


学習器clfの実態はパイプラインですから、<b>これを学習済みモデルとして保存しておけば、学習済みscl、学習済みreduct(PCA)、学習済みclf（モデル）の3つが学習状態で保存（永続化）されます</b>。

<b>[確認してみよう]</b> 学習済みモデルの永続化の方法をScikit-learnの公式ドキュメントで調査しよう。
- https://scikit-learn.org/stable/modules/model_persistence.html <br>
<b>[確認してみよう]</b> 特徴量選択のクラスにRFECVがあります。RFEと何が違うかScikit-learnの公式ドキュメントで調べてみよう。
- https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFECV.html

# スコアリングフェーズのデータ処理

ローン審査データを使って<b>モデリング段階のデータ処理フローをおさらいし、</b><br>その後、スコアリング段階のデータ処理で必要となるテクニックを学びましょう。

In [20]:
# import sample data: Loan screening data for classification 
import pandas as pd
df = pd.read_csv('../python/data/av_loan_u6lujuX_CVtuZ9i.csv')
# df.head()
X =  df.iloc[:,:-1]
ID =  X.iloc[:,[0]]
X = X.drop('Loan_ID',axis=1)
y =  df.iloc[:,-1]

# check the shape
print('--------------------------------------')
print('Raw shape: (%i,%i)' %df.shape)
print('X shape: (%i,%i)' %X.shape)

# ローン審査でNOとなったサンプルを1（正例）として変換
class_mapping = {'N':1,'Y':0} #  Nが1であることに注意する。
y =  y.map(class_mapping)
print('---------------------------------------')
print(y.value_counts())
print('---------------------------------------')
print(ID.join(X).join(y).dtypes)
ID.join(X).join(y).head()

# 表示列数のオプション変更
pd.options.display.max_columns = 50

--------------------------------------
Raw shape: (614,13)
X shape: (614,11)
---------------------------------------
0    422
1    192
Name: Loan_Status, dtype: int64
---------------------------------------
Loan_ID               object
Gender                object
Married               object
Dependents            object
Education             object
Self_Employed         object
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area         object
Loan_Status            int64
dtype: object


##### モデル用データの前処理：カテゴリ変数の数量化と欠損対応
まず、カテゴリ変数のone-hotエンコーディングを行います。

In [24]:
ohe_columns = ['Dependents',
               'Gender',
               'Married',
               'Education',
               'Self_Employed',
               'Property_Area']

X_ohe =  pd.get_dummies(X,dummy_na=True,columns= ohe_columns)
print('X_ohe shape(%i,%i)' % X_ohe.shape)
display(X_ohe.head())

X_ohe shape(614,26)


ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History  Dependents_0  Dependents_1  Dependents_2  Dependents_3+  \
0             1.0             1             0             0              0   
1             1.0             0             1             0              0   
2             1.0             1             0             0              0   
3             1.0             1             0             0              0   
4             1.0             1             0             0              0   

   Dependents_nan  Gender_Female  Gender_Male  Gender_nan  Married_No  \
0               0              0            1           0           1   
1               0              0            1           0           0   
2               0              0            1           0           0   
3               0              0            1           0           0   
4               0              0            1           0           1   

   Married_Yes  Married_nan  Education_Graduate  Education_Not Graduate  \
0            0            0                   1                       0   
1            1            0                   1                       0   
2            1            0                   1                       0   
3            1            0                   0                       1   
4            0            0                   1                       0   

   Education_nan  Self_Employed_No  Self_Employed_Yes  Self_Employed_nan  \
0              0                 1                  0                  0   
1              0                 1                  0                  0   
2              0                 0                  1                  0   
3              0                 1                  0                  0   
4              0                 1                  0                  0   

   Property_Area_Rural  Property_Area_Semiurban  Property_Area_Urban  \
0                    0                        0                    1   
1                    1                        0                    0   
2                    0                        0                    1   
3                    0                        0                    1   
4                    0                        0                    1   

   Property_Area_nan  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0

##### モデル用データの前処理：数値変数の欠損対応
次に、数値変数の欠損値を各列平均値で置換します。

In [31]:
from sklearn.impute import SimpleImputer

# 欠損値NaNを平均値(mean)で置換
imp = SimpleImputer()
imp.fit(X_ohe)
# 学習済みImputerを適用しX_newの欠損値を置換
X_ohe_columns =  X_ohe.columns.values
X_ohe = pd.DataFrame(imp.transform(X_ohe),columns=X_ohe_columns)

# 結果表示
display(X_ohe.head())

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0           5849.0                0.0  146.412162             360.0   
1           4583.0             1508.0  128.000000             360.0   
2           3000.0                0.0   66.000000             360.0   
3           2583.0             2358.0  120.000000             360.0   
4           6000.0                0.0  141.000000             360.0   

   Credit_History  Dependents_0  Dependents_1  Dependents_2  Dependents_3+  \
0             1.0           1.0           0.0           0.0            0.0   
1             1.0           0.0           1.0           0.0            0.0   
2             1.0           1.0           0.0           0.0            0.0   
3             1.0           1.0           0.0           0.0            0.0   
4             1.0           1.0           0.0           0.0            0.0   

   Dependents_nan  Gender_Female  Gender_Male  Gender_nan  Married_No  \
0             0.0            0.0          1.0         0.0         1.0   
1             0.0            0.0          1.0         0.0         0.0   
2             0.0            0.0          1.0         0.0         0.0   
3             0.0            0.0          1.0         0.0         0.0   
4             0.0            0.0          1.0         0.0         1.0   

   Married_Yes  Married_nan  Education_Graduate  Education_Not Graduate  \
0          0.0          0.0                 1.0                     0.0   
1          1.0          0.0                 1.0                     0.0   
2          1.0          0.0                 1.0                     0.0   
3          1.0          0.0                 0.0                     1.0   
4          0.0          0.0                 1.0                     0.0   

   Education_nan  Self_Employed_No  Self_Employed_Yes  Self_Employed_nan  \
0            0.0               1.0                0.0                0.0   
1            0.0               1.0                0.0                0.0   
2            0.0               0.0                1.0                0.0   
3            0.0               1.0                0.0                0.0   
4            0.0               1.0                0.0                0.0   

   Property_Area_Rural  Property_Area_Semiurban  Property_Area_Urban  \
0                  0.0                      0.0                  1.0   
1                  1.0                      0.0                  0.0   
2                  0.0                      0.0                  1.0   
3                  0.0                      0.0                  1.0   
4                  0.0                      0.0                  1.0   

   Property_Area_nan  
0                0.0  
1                0.0  
2                0.0  
3                0.0  
4                0.0

##### モデル用データの前処理：次元圧縮（特徴選択）
続けて、特徴次元の圧縮を図ります。

In [34]:
from sklearn.feature_selection import RFE
from sklearn.ensemble import RandomForestClassifier
selector =  RFE(RandomForestClassifier(n_estimators=100,random_state=1),
               n_features_to_select=10,
               step=.05)
selector.fit(X_ohe,y)

X_fin = pd.DataFrame(selector.transform(X_ohe),columns= X_ohe_columns[selector.support_])
print('X_fin shape:(%i,%i)' % X_fin.shape)
X_fin.head()

X_fin shape:(614,10)


ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0           5849.0                0.0  146.412162             360.0   
1           4583.0             1508.0  128.000000             360.0   
2           3000.0                0.0   66.000000             360.0   
3           2583.0             2358.0  120.000000             360.0   
4           6000.0                0.0  141.000000             360.0   

   Credit_History  Dependents_0  Married_No  Education_Graduate  \
0             1.0           1.0         1.0                 1.0   
1             1.0           0.0         0.0                 1.0   
2             1.0           1.0         0.0                 1.0   
3             1.0           1.0         0.0                 0.0   
4             1.0           1.0         1.0                 1.0   

   Property_Area_Rural  Property_Area_Semiurban  
0                  0.0                      0.0  
1                  1.0                      0.0  
2                  0.0                      0.0  
3                  0.0                      0.0  
4                  0.0                      0.0

##### スコア用データの前処理
さて、スコア用データの前処理です。本処理は以下要件を満たす必要があります。
- 上記10次元の特徴量をこの並びの通りに変換する必要
- 学習済みSimpleImputerインスタンスによる欠損値補完
- 学習済みRFEインスタンスによる特徴量選択

なぜなら、以下理由のためです。
- モデル用データとスコア用データの並び順が違うことを学習済みモデルはわからない
- 並び順が異なると、学習済みSimpleImputerも学習済みRFEもインデックス情報が狂う
- 数値変数の欠損値は「学習段階」の平均値でしか置換できない（未来情報は使えない）
- 何が重要な変数かは「学習段階」のXとyの関係からしか知りようがない（同上）

In [45]:
# import sample data for classificatio
df_s = pd.read_csv('../python/data/av_loan_test_Y3wMUE5_7gLdaTN.csv', header=0)
ID_s = df_s.iloc[:,[0]]            # 第0列はPK（Loan_ID）なのでIDとしてセット
X_s  = df_s.drop('Loan_ID',axis=1) # Loan_IDはID情報なので特徴ベクトルから削除

# check the shape
print('Raw shape: (%i,%i)' %df_s.shape)
print('X shape: (%i,%i)' %X_s.shape)
print('-------------------------------')
print(X_s.dtypes)

Raw shape: (333,12)
X shape: (333,11)
-------------------------------
Gender                object
Married               object
Dependents           float64
Education             object
Self_Employed         object
ApplicantIncome        int64
CoapplicantIncome      int64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area         object
dtype: object


##### スコア用データの前処理：カテゴリ変数の数量化と欠損対応
本処理はスコア用データに対し、モデル用データとは独立に実施します。

In [46]:
X_ohe_s = pd.get_dummies(X_s,
                         dummy_na=True,
                         columns=ohe_columns)
print('X_ohe_s shape:(%i,%i)' % X_ohe_s.shape)
X_ohe_s.head()

X_ohe_s shape:(333,26)


ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5720                  0       110.0             360.0   
1             3076               1500       126.0             360.0   
2             5000               1800       208.0             360.0   
3             2340               2546       100.0             360.0   
4             3276                  0        78.0             360.0   

   Credit_History  Dependents_0.0  Dependents_1.0  Dependents_2.0  \
0             1.0               1               0               0   
1             1.0               0               1               0   
2             1.0               0               0               1   
3             NaN               0               0               1   
4             1.0               1               0               0   

   Dependents_nan  Gender_Female  Gender_Male  Gender_Unknown  Gender_nan  \
0               0              0            1               0           0   
1               0              0            1               0           0   
2               0              0            1               0           0   
3               0              0            1               0           0   
4               0              0            1               0           0   

   Married_No  Married_Yes  Married_nan  Education_Graduate  \
0           0            1            0                   1   
1           0            1            0                   1   
2           0            1            0                   1   
3           0            1            0                   1   
4           1            0            0                   0   

   Education_Not Graduate  Education_nan  Self_Employed_No  Self_Employed_Yes  \
0                       0              0                 1                  0   
1                       0              0                 1                  0   
2                       0              0                 1                  0   
3                       0              0                 1                  0   
4                       1              0                 1                  0   

   Self_Employed_nan  Property_Area_Rural  Property_Area_Semiurban  \
0                  0                    0                        0   
1                  0                    0                        0   
2                  0                    0                        0   
3                  0                    0                        0   
4                  0                    0                        0   

   Property_Area_Urban  Property_Area_nan  
0                    1                  0  
1                    1                  0  
2                    1                  0  
3                    1                  0  
4                    1                  0

##### スコア用データの前処理：one-hotエンコーディング後のデータ整合チェック
さて、one-hotエンコーディング後のモデル用とスコア用データの整合性を確認してみます。
確認のため、Pythonのset型（集合型変数）を利用しています。

In [52]:
# pythonのset関数で一つしかない値をだす。
# 洗濯した特徴量がどれだけ同じか
cols_model = set(X_ohe.columns.values)
cols_score = set(X_ohe_s.columns.values)

diff1 =  cols_model - cols_score
print('Modelのみ:%s'% diff1)

diff2 = cols_score - cols_model
print('Scoreのみ:%s' % diff2)
# このように型が変わっているので、明示的に型を設定することで整合性を保つ。

Modelのみ:{'Dependents_1', 'Dependents_0', 'Dependents_2', 'Dependents_3+'}
Scoreのみ:{'Dependents_0.0', 'Dependents_1.0', 'Gender_Unknown', 'Dependents_2.0'}


実は、このスコア用データは、以下２つの細工が施されていました。
1. Gender変数に"Unknown"というカテゴリ値を新しく追加
2. Dependents変数の"3+"というカテゴリ値を除外（残された値は0,1,2の3種類）

結果としてモデル用とスコア用のデータ間で、以下不整合の生じる可能性があるとわかります。
1. モデルデータにないカラムが生成される可能性（Gender_Unknown)
1. モデルデータにあったカラムが消える可能性（Dependents_3+）
1. データ型の違いが理由で①/②が生じる可能性

### スコアリングフェーズにおけるデータ処理（解決編）

##### 課題把握編のおさらい
前編にて、スコアリングデータのone-hotエンコーディング後に、以下の問題が生じる得ることを確認しました。<br>
そこでこの問題が生じないよう、コードを補強していくことにしましょう。
- モデルデータにないカラムが生成される可能性
- モデルデータにあったカラムが消える可能性
- データ型の違いが理由で上記を生じさせてしまう可能性

##### データ前処理時の前提
- 私達はどのデータが数値変数で、どのデータがカテゴリ変数かを事前に把握している
- そして今回のカテゴリ変数は以下6変数とする
 - Dependents, Gender, Married, Education, Self_Employed, Property_Area

##### モデル用データの前処理
最初の補強は、ファイル読込段階で<b>カテゴリ変数をobject型として明記すること</b>です。

### 型が変わっているので、明示的に型を設定することで整合性を保つ。

In [55]:
# カテゴリ変数をリストで設定
ohe_columns = ['Dependents',
#                'Gender',
               'Married',
               'Education',
               'Self_Employed',
               'Property_Area']

# カテゴリ変数をobject型で読み込むための準備
my_dtype = {'Dependents':object,
            'Gender':object,
            'Married':object,
            'Education':object,
            'Self_Employed':object,
            'Property_Area':object}

# 表示オプションの変更
# import pandas as pd
# pd.options.display.max_columns = 50

In [61]:
# 型を統一して実行する。
# モデル用
# Loan screening model data for classification 
df = pd.read_csv('../python/data/av_loan_u6lujuX_CVtuZ9i.csv',
                 header=0,
                 dtype=my_dtype)
# テスト用
# Loan screening test data for classification 
df_s = pd.read_csv('../python/data/av_loan_test_Y3wMUE5_7gLdaTN.csv',
                   header=0,
                   dtype=my_dtype)

In [64]:
 # 上記と全く同じコード
    
# カテゴリ変数をリストで設定
ohe_columns = ['Dependents',
               'Gender',
               'Married',
               'Education',
               'Self_Employed',
               'Property_Area']

# カテゴリ変数をobject型で読み込むための準備
my_dtype = {'Dependents':object,
            'Gender':object,
            'Married':object,
            'Education':object,
            'Self_Employed':object,
            'Property_Area':object}

# 表示オプションの変更
import pandas as pd
pd.options.display.max_columns = 50
# Loan screening model data for classification 
# df = pd.read_csv('./data/av_loan_u6lujuX_CVtuZ9i.csv',
#                  header=0,
#                  dtype=my_dtype)

X  = df.iloc[:,:-1]            # 最終列以前を特徴量
ID = X.iloc[:,[0]]             # 第0列はPK（Loan_ID）なのでIDとしてセット
X  = X.drop('Loan_ID', axis=1) # Loan_IDは特徴ベクトルから削除
y  = df.iloc[:,-1]             # 最終列を正解データ

# check the shape
print('---------------------------------')
print('Raw shape: (%i,%i)' %df.shape)
print('X shape: (%i,%i)' %X.shape)
print('---------------------------------')
print(X.dtypes)

# ローン審査でNOとなったサンプルを1（正例）として変換
class_mapping = {'N':1, 'Y':0}
y = y.map(class_mapping)
X_ohe = pd.get_dummies(X,
                       dummy_na=True,
                       columns=ohe_columns)

print('X_ohe shape:(%i,%i)' % X_ohe.shape)
X_ohe.head()
from sklearn.impute import SimpleImputer

imp = SimpleImputer()
imp.fit(X_ohe)

X_ohe_columns = X_ohe.columns.values
X_ohe = pd.DataFrame(imp.transform(X_ohe), columns=X_ohe_columns)

X_ohe.head()
from sklearn.feature_selection import RFE
from sklearn.ensemble import RandomForestClassifier

selector = RFE(RandomForestClassifier(n_estimators=100, random_state=1),
               n_features_to_select=10,
               step=.05)

selector.fit(X_ohe,y)

X_fin = X_ohe.loc[:, X_ohe_columns[selector.support_]]
print('X_fin shape:(%i,%i)' % X_fin.shape)
X_fin.head()
# df_s = pd.read_csv('./data/av_loan_test_Y3wMUE5_7gLdaTN.csv',
#                    header=0,
#                    dtype=my_dtype)

ID_s = df_s.iloc[:,[0]]            # 第0列はPK（Loan_ID）なのでIDとしてセット
X_s  = df_s.drop('Loan_ID',axis=1) # Loan_IDは特徴ベクトルから削除

# check the shape
print('---------------------------------')
print('Raw shape: (%i,%i)' %df_s.shape)
print('X shape: (%i,%i)' %X_s.shape)
print('---------------------------------')
print(X_s.dtypes)
ID_s.join(X_s).head(3)
X_ohe_s = pd.get_dummies(X_s,
                         dummy_na=True,
                         columns=ohe_columns)
print('X_ohe_s shape:(%i,%i)' % X_ohe_s.shape)
X_ohe_s.head(3)

---------------------------------
Raw shape: (614,13)
X shape: (614,11)
---------------------------------
Gender                object
Married               object
Dependents            object
Education             object
Self_Employed         object
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area         object
dtype: object
X_ohe shape:(614,26)
X_fin shape:(614,10)
---------------------------------
Raw shape: (333,12)
X shape: (333,11)
---------------------------------
Gender                object
Married               object
Dependents            object
Education             object
Self_Employed         object
ApplicantIncome        int64
CoapplicantIncome      int64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area         object
dtype: object
X_ohe_s shape:(333,26)


ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5720                  0       110.0             360.0   
1             3076               1500       126.0             360.0   
2             5000               1800       208.0             360.0   

   Credit_History  Dependents_0  Dependents_1  Dependents_2  Dependents_nan  \
0             1.0             1             0             0               0   
1             1.0             0             1             0               0   
2             1.0             0             0             1               0   

   Gender_Female  Gender_Male  Gender_Unknown  Gender_nan  Married_No  \
0              0            1               0           0           0   
1              0            1               0           0           0   
2              0            1               0           0           0   

   Married_Yes  Married_nan  Education_Graduate  Education_Not Graduate  \
0            1            0                   1                       0   
1            1            0                   1                       0   
2            1            0                   1                       0   

   Education_nan  Self_Employed_No  Self_Employed_Yes  Self_Employed_nan  \
0              0                 1                  0                  0   
1              0                 1                  0                  0   
2              0                 1                  0                  0   

   Property_Area_Rural  Property_Area_Semiurban  Property_Area_Urban  \
0                    0                        0                    1   
1                    0                        0                    1   
2                    0                        0                    1   

   Property_Area_nan  
0                  0  
1                  0  
2                  0

In [63]:
cols_model = set(X_ohe.columns.values)
cols_score = set(X_ohe_s.columns.values)

diff1 = cols_model - cols_score
print('モデルのみに存在する項目: %s' % diff1)

diff2 = cols_score - cols_model
print('スコアのみに存在する項目: %s' % diff2)

モデルのみに存在する項目: {'Dependents_3+'}
スコアのみに存在する項目: {'Gender_Unknown'}


##### 残されたデータ不整合
データ型の違いによる不一致は消え、以下の違いだけが残りました。
- Dependents_3+はスコアリングデータには存在しない
- Gender_Unknownはスコアリングデータで新たに登場した

##### 不整合解消のための基本指針
以下が基本方針となります。暗記ではなく、なぜそうすべきなのかを考えてみましょう。
- モデル用にはあるが、スコア用に存在しない変数は復活させる
- スコア用データにあるが、モデル用に存在しない変数は削除する

##### 不整合解消のためのtips：Pandasデータフレームのconcat関数
上記指針実現の一つの方法が、データフレームを縦に結合するconcat処理です。<br>一旦サンプルデータからはなれ、concat関数の動作を確認しましょう。

In [77]:
# サンプル
# カラム構成が同じデータフレームの結合
df1 = pd.DataFrame([[1,2,3]], columns=['c1','c2','c3'])
df2 = pd.DataFrame([[3,2,1]], columns=['c1','c2','c3'])
df_all = pd.concat([df1, df2])
df_all

c1  c2  c3
0   1   2   3
0   3   2   1

In [78]:
# サンプル
# カラム構成が異なるデータフレームの結合
df3 = pd.DataFrame([[0,1,2,3]],columns=['c0','c1','c3','c4'])
df_all = pd.concat([df_all, df3])
df_all
# df3

/Users/humu/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  after removing the cwd from sys.path.


c0  c1   c2  c3   c4
0  NaN   1  2.0   3  NaN
0  NaN   3  2.0   1  NaN
0  0.0   1  NaN   2  3.0

##### concat関数の動作
- c0とc4は、元々のデータフレーム（df_all）に存在しなかった変数なので、スコア用データに初めて登場した変数と言えます。
 - よって、対応は削除となります。
- 一方、c2はモデル用データにはあったがスコア用データにはなかった変数と言えます。
 - よって、対応は変数の復活です。
 - かつ、このような不一致はカテゴリ変数でのみ生じるので、ゼロ補完が妥当です。

##### スコア用データの前処理：one-hotエンコーディング後のデータ不整合の解消
それではサンプルデータに戻ります。<br>
モデリング時点のone-hotエンコーディング後のカラム構成は`X_ohe_columns`でした。<br>
そこで、このカラム構成だけを持った（データ部分は持たない）データフレームを作ります。

In [69]:
df_cols_m = pd.DataFrame(None,
                         columns=X_ohe_columns,
                         dtype=float)
display(df_cols_m)

Empty DataFrame
Columns: [ApplicantIncome, CoapplicantIncome, LoanAmount, Loan_Amount_Term, Credit_History, Dependents_0, Dependents_1, Dependents_2, Dependents_3+, Dependents_nan, Gender_Female, Gender_Male, Gender_nan, Married_No, Married_Yes, Married_nan, Education_Graduate, Education_Not Graduate, Education_nan, Self_Employed_No, Self_Employed_Yes, Self_Employed_nan, Property_Area_Rural, Property_Area_Semiurban, Property_Area_Urban, Property_Area_nan]
Index: []

上記に対し、スコアリング時点のone-hotエンコーディング後のデータを結合します。<br>
実行し以下を確認しましょう。
- `Dependents_3+`が復活（ただし、値は欠損状態）
- `Gender_Unknown`はまだ存在

In [70]:
X_ohe_s2 = pd.concat([df_cols_m, X_ohe_s])
print(X_ohe_s2.shape)
display(X_ohe_s2.head(3))

(333, 27)


/Users/humu/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """Entry point for launching an IPython kernel.


ApplicantIncome  CoapplicantIncome  Credit_History  Dependents_0  \
0           5720.0                0.0             1.0           1.0   
1           3076.0             1500.0             1.0           0.0   
2           5000.0             1800.0             1.0           0.0   

   Dependents_1  Dependents_2  Dependents_3+  Dependents_nan  \
0           0.0           0.0            NaN             0.0   
1           1.0           0.0            NaN             0.0   
2           0.0           1.0            NaN             0.0   

   Education_Graduate  Education_Not Graduate  Education_nan  Gender_Female  \
0                 1.0                     0.0            0.0            0.0   
1                 1.0                     0.0            0.0            0.0   
2                 1.0                     0.0            0.0            0.0   

   Gender_Male  Gender_Unknown  Gender_nan  LoanAmount  Loan_Amount_Term  \
0          1.0             0.0         0.0       110.0             360.0   
1          1.0             0.0         0.0       126.0             360.0   
2          1.0             0.0         0.0       208.0             360.0   

   Married_No  Married_Yes  Married_nan  Property_Area_Rural  \
0         0.0          1.0          0.0                  0.0   
1         0.0          1.0          0.0                  0.0   
2         0.0          1.0          0.0                  0.0   

   Property_Area_Semiurban  Property_Area_Urban  Property_Area_nan  \
0                      0.0                  1.0                0.0   
1                      0.0                  1.0                0.0   
2                      0.0                  1.0                0.0   

   Self_Employed_No  Self_Employed_Yes  Self_Employed_nan  
0               1.0                0.0                0.0  
1               1.0                0.0                0.0  
2               1.0                0.0                0.0

In [79]:
set_Xm = set(X_ohe.columns.values)
set_Xs = set(X_ohe_s.columns.values)
print(set_Xm,set_Xs)

{'Dependents_3+', 'Gender_Female', 'Self_Employed_Yes', 'Education_Graduate', 'Dependents_0', 'Married_nan', 'Gender_Male', 'Property_Area_Rural', 'Education_nan', 'Education_Not Graduate', 'Property_Area_nan', 'CoapplicantIncome', 'Property_Area_Urban', 'Gender_nan', 'Credit_History', 'Dependents_1', 'Self_Employed_nan', 'Dependents_nan', 'Married_No', 'Married_Yes', 'Property_Area_Semiurban', 'LoanAmount', 'Dependents_2', 'Loan_Amount_Term', 'ApplicantIncome', 'Self_Employed_No'} {'Gender_Female', 'Self_Employed_Yes', 'Education_Graduate', 'Dependents_0', 'Married_nan', 'Gender_Male', 'Gender_Unknown', 'Property_Area_Rural', 'Education_nan', 'Education_Not Graduate', 'Property_Area_nan', 'CoapplicantIncome', 'Property_Area_Urban', 'Gender_nan', 'Credit_History', 'Dependents_1', 'Self_Employed_nan', 'Dependents_nan', 'Married_No', 'Married_Yes', 'Property_Area_Semiurban', 'LoanAmount', 'Dependents_2', 'Loan_Amount_Term', 'ApplicantIncome', 'Self_Employed_No'}


In [81]:
# 次に、スコアリングデータのみに登場する変数`Gender_Unknown`を削除しましょう。
X_ohe_s3 = X_ohe_s2.drop(list(set_Xs-set_Xm),axis=1) # スコアようにだけ存在する特徴量の削除

print(X_ohe_s3.shape)
display(X_ohe_s3.head(3))

(333, 26)


ApplicantIncome  CoapplicantIncome  Credit_History  Dependents_0  \
0           5720.0                0.0             1.0           1.0   
1           3076.0             1500.0             1.0           0.0   
2           5000.0             1800.0             1.0           0.0   

   Dependents_1  Dependents_2  Dependents_3+  Dependents_nan  \
0           0.0           0.0            NaN             0.0   
1           1.0           0.0            NaN             0.0   
2           0.0           1.0            NaN             0.0   

   Education_Graduate  Education_Not Graduate  Education_nan  Gender_Female  \
0                 1.0                     0.0            0.0            0.0   
1                 1.0                     0.0            0.0            0.0   
2                 1.0                     0.0            0.0            0.0   

   Gender_Male  Gender_nan  LoanAmount  Loan_Amount_Term  Married_No  \
0          1.0         0.0       110.0             360.0         0.0   
1          1.0         0.0       126.0             360.0         0.0   
2          1.0         0.0       208.0             360.0         0.0   

   Married_Yes  Married_nan  Property_Area_Rural  Property_Area_Semiurban  \
0          1.0          0.0                  0.0                      0.0   
1          1.0          0.0                  0.0                      0.0   
2          1.0          0.0                  0.0                      0.0   

   Property_Area_Urban  Property_Area_nan  Self_Employed_No  \
0                  1.0                0.0               1.0   
1                  1.0                0.0               1.0   
2                  1.0                0.0               1.0   

   Self_Employed_Yes  Self_Employed_nan  
0                0.0                0.0  
1                0.0                0.0  
2                0.0                0.0

次に、スコアリングでは登場しなかったデータ項目をゼロ埋めします。<br>
実行し以下を確認しましょう。
- 変数`Depend_3+`がゼロで埋まっていること

In [83]:
# nan値を0で埋める。
X_ohe_s3.loc[:,list(set_Xm-set_Xs)] = X_ohe_s3.loc[:,list(set_Xm-set_Xs)].fillna(0,axis=1)
X_ohe_s3.head(3)

ApplicantIncome  CoapplicantIncome  Credit_History  Dependents_0  \
0           5720.0                0.0             1.0           1.0   
1           3076.0             1500.0             1.0           0.0   
2           5000.0             1800.0             1.0           0.0   

   Dependents_1  Dependents_2  Dependents_3+  Dependents_nan  \
0           0.0           0.0            0.0             0.0   
1           1.0           0.0            0.0             0.0   
2           0.0           1.0            0.0             0.0   

   Education_Graduate  Education_Not Graduate  Education_nan  Gender_Female  \
0                 1.0                     0.0            0.0            0.0   
1                 1.0                     0.0            0.0            0.0   
2                 1.0                     0.0            0.0            0.0   

   Gender_Male  Gender_nan  LoanAmount  Loan_Amount_Term  Married_No  \
0          1.0         0.0       110.0             360.0         0.0   
1          1.0         0.0       126.0             360.0         0.0   
2          1.0         0.0       208.0             360.0         0.0   

   Married_Yes  Married_nan  Property_Area_Rural  Property_Area_Semiurban  \
0          1.0          0.0                  0.0                      0.0   
1          1.0          0.0                  0.0                      0.0   
2          1.0          0.0                  0.0                      0.0   

   Property_Area_Urban  Property_Area_nan  Self_Employed_No  \
0                  1.0                0.0               1.0   
1                  1.0                0.0               1.0   
2                  1.0                0.0               1.0   

   Self_Employed_Yes  Self_Employed_nan  
0                0.0                0.0  
1                0.0                0.0  
2                0.0                0.0

##### スコア用データの前処理：スコア用データの変数の並び順の制御
最後に、モデリング時点のデータ項目の並び順を明示的に担保しましょう。<br>
以下の通り、`reindex`関数を使うことで並び順を制御できます。

In [86]:
test = pd.DataFrame([[1,2,3]], columns=['c1','c2','c3'])
display(test)
test = test.reindex(['c2','c3','c1'], axis=1) # columnsの順番を変更
display(test)

c1  c2  c3
0   1   2   3

c2  c3  c1
0   2   3   1

ここでは、モデリング時点のone-hotエンコーディング後の並び順に制御します。

In [87]:
X_ohe_s3 = X_ohe_s3.reindex(X_ohe.columns.values,axis=1)
X_ohe_s3.head(3)

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0           5720.0                0.0       110.0             360.0   
1           3076.0             1500.0       126.0             360.0   
2           5000.0             1800.0       208.0             360.0   

   Credit_History  Dependents_0  Dependents_1  Dependents_2  Dependents_3+  \
0             1.0           1.0           0.0           0.0            0.0   
1             1.0           0.0           1.0           0.0            0.0   
2             1.0           0.0           0.0           1.0            0.0   

   Dependents_nan  Gender_Female  Gender_Male  Gender_nan  Married_No  \
0             0.0            0.0          1.0         0.0         0.0   
1             0.0            0.0          1.0         0.0         0.0   
2             0.0            0.0          1.0         0.0         0.0   

   Married_Yes  Married_nan  Education_Graduate  Education_Not Graduate  \
0          1.0          0.0                 1.0                     0.0   
1          1.0          0.0                 1.0                     0.0   
2          1.0          0.0                 1.0                     0.0   

   Education_nan  Self_Employed_No  Self_Employed_Yes  Self_Employed_nan  \
0            0.0               1.0                0.0                0.0   
1            0.0               1.0                0.0                0.0   
2            0.0               1.0                0.0                0.0   

   Property_Area_Rural  Property_Area_Semiurban  Property_Area_Urban  \
0                  0.0                      0.0                  1.0   
1                  0.0                      0.0                  1.0   
2                  0.0                      0.0                  1.0   

   Property_Area_nan  
0                0.0  
1                0.0  
2                0.0

##### スコア用データの前処理：数値変数の欠損値対応
ここまで整合させてようやく、学習済みSimpleImputerインスタンスが適用可能となります。

並び方がSimpleImputerで学習させた時の順番でないと<br>
別の特徴量の平均が入ることがあるので、<br>
いらないカラムをのぞいて順番を合わせて(reindex関数)から適応する。<br>
- dummies
- X_new
- imp.fit(X_new)
の順番でしたので、X_newに順番を合わせる。

In [91]:
print('欠損個数（数値変数の欠損補完前）',X_ohe_s3.isnull().sum().sum())
X_ohe_s4 = pd.DataFrame(imp.transform(X_ohe_s3),columns=X_ohe_columns)
print('欠損個数（数値変数の欠損補完後）',X_ohe_s4.isnull().sum().sum())

欠損個数（数値変数の欠損補完前） 34
欠損個数（数値変数の欠損補完後） 0


RFEによって選択された変数の位置はsupport_属性から取得できたので、<br>スコアリングデータの特徴量の最終形は以下のようになります。

In [92]:
X_fin_s = X_ohe_s4.loc[:, X_ohe_columns[selector.support_]]
print(X_fin_s.shape)
X_fin_s.head(3)

(333, 10)


ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0           5720.0                0.0       110.0             360.0   
1           3076.0             1500.0       126.0             360.0   
2           5000.0             1800.0       208.0             360.0   

   Credit_History  Dependents_0  Married_No  Education_Graduate  \
0             1.0           1.0         0.0                 1.0   
1             1.0           0.0         0.0                 1.0   
2             1.0           0.0         0.0                 1.0   

   Property_Area_Rural  Property_Area_Semiurban  
0                  0.0                      0.0  
1                  0.0                      0.0  
2                  0.0                      0.0